### ISSUE TITLE
Separating Material and Cupertino from Flutter

### ISSUE THREAD

**Issue by Levi-Lesches** [flutter/flutter#111234]
_Originally posted in https://github.com/flutter/flutter/issues/91605#issuecomment-1154573877_

----

I know this part of the issue is an old topic, but I'd like to contribute my thoughts. I agree that the Material and Cupertino widgets should be separated from the main Flutter repo (and included by default in the `pubspec.yaml` from `flutter create`) for several reasons: 
- It enables Flutter and Material to have different views on breaking changes. For example, Flutter seems to be very against breaking changes, favoring a time-based deprecation scheme instead of semantic versioning. Dart also uses its own language versioning process. A separate package would allow the widgets to stay in sync with the Material spec, independent of Flutter's policies (not that I disagree with those policies).
- On the same note, if a dev prefers, they can mix and match versions of Flutter and Material (to an extent, probably). For example, if a dev was on Flutter 2, they could easily make the purely technical switch to Flutter 3 and gain all the performance improvements without committing to the stylistic changes of Material 3.
- This platform-agnostic separation would make using other platforms feel less hacky. It's no doubt that Flutter intends for Material to be the "default", and for documentation purposes, it still could be. Have all the codelabs follow Material, sure. But today, adding support for other platforms is not straightforward. 
  - Some widgets have a `.adaptive` constructor, like [`Switch.adaptive`](https://api.flutter.dev/flutter/material/Switch/Switch.adaptive.html). It is unclear which platforms these constructors support since they are included in the `material` library. Also, these widgets can't take advantage of other supported platforms like `fluent_ui` since that's a separate package. 
  - Other widgets have platform adaptive behavior, like `Scrollable`. When over-scrolled, it sends [`GlowingOverscrollIndicator`](https://api.flutter.dev/flutter/widgets/GlowingOverscrollIndicator-class.html) or [`BouncingScrollPhysics`](https://api.flutter.dev/flutter/widgets/StretchingOverscrollIndicator-class.html) notifications, depending on the platform. It can even send [`StretchingOverscrollIndicator`](https://api.flutter.dev/flutter/widgets/StretchingOverscrollIndicator-class.html) notifications like Material 3. Again, it's unclear which behavior will be chosen. Again, it is also impossible for these widgets to function correctly on platforms with their own package, like `fluent_ui`. 
  - Other widgets are split between `material`, `cupertino`, `fluent_ui`, and other libraries and must be chosen manually. Given the range of behaviors listed above, it's unclear when to rely on the default Material widget or find an alternative in these massive libraries. As mentioned, some even depend on others, making them not truly independent. Also, the alternative widgets don't feel official since they're prefixed by platform. Between `CupertinoTheme` and `Theme`, it feels like `Theme` is the "correct" version to use. 
- This will make it easier for developers to introduce their own globally themed widgets. Separately from using `ThemeData`, there are cases where one wants every or most Material widget to be replaced with their own. A modular framework allows devs to create their own widgets, hardly interacting with the default Material library. One example would be when porting Flutter to a new platform. Another example is a wiki site for a game that wants to have buttons, text, and displays themed like the game's UI. For an example, see [this page](https://www.factorio.com/blog/) from Factorio's website. Imagine importing `package:factorio` instead of `material` to create that site. 

TL;DR: Separating platform widgets into their own package is useful for breaking changes, but it primarily declares an intent to support a given platform's conventions. While this doesn't have to replace the "out-of-the-box" support for a reasonable default like Material, it does allow each platform to take its own spin on Flutter. 

----

Then there's the problem of easily supporting different platforms' conventions by default, as Flutter currently does. I'm not sure the _best_ way to do this, but from my last bullet, one can make an umbrella package that automatically chooses the right widget for the user's platform, and users would import that. So imagine a `mobile_ui` package with its own `AlertDialog` that chooses between a Material dialog on Android or a `CupertinoDialog` on iOS.

**Comment by stuartmorgan** [comment]
This is a duplicate of https://github.com/flutter/flutter/issues/101479; closing to consolidate discussion there.

**Comment by Levi-Lesches** [comment]
Thanks, I was just in the process of posting there too.
